File: 백준/Gold/2493. 탑/탑.js
const fs = require("fs");
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

let num = Number(input[0]);
let top_arr = input[1].split(" ").map(Number);

// 0 1 2 3 4
// 6 9 5 7 4

const answer = Array(num).fill(0);
const stack = [];

for (let i = 0; i < num; i++) {
    while (stack.length > 0 && top_arr[stack[stack.length - 1]] < top_arr[i]) {
        stack.pop();
    }
    if (stack.length > 0) {
        answer[i] = stack[stack.length - 1] + 1; 
    } 
    stack.push(i);
}

console.log(answer.join(" "));

File: 백준/Gold/2493. 탑/탑.py
import sys
input = sys.stdin.readline

n = int(input())
heights = list(map(int,input().split()))

stack = []
answer = [0] * n
for i in range(n):
    # 왼쪽에 값이 있고 자신보다 클 때 수신 받음
    # 나보다 작은 탑은 제거
    while stack and heights[stack[-1]] < heights[i]:
        stack.pop()

    # 스택에 남아있는 맨 위 = 나보다 크거나 같은 가장 가까운 탑
    if stack:
        answer[i] = stack[-1] + 1 # 인덱스 계산

    stack.append(i)

print(' '.join(map(str, answer)))